File: app-android/src/main/java/com/ustadmobile/port/android/view/ClazzListRecyclerAdapter.java
package com.ustadmobile.port.android.view;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.ClazzListPresenter;
import com.ustadmobile.lib.db.entities.ClazzWithNumStudents;

/**
 * The ClazzList Recycler Adapter used here.
 */
public class ClazzListRecyclerAdapter extends
        PagedListAdapter<ClazzWithNumStudents, ClazzListRecyclerAdapter.ClazzViewHolder> {

    Context theContext;
    private Fragment theFragment;
    private ClazzListPresenter thePresenter;

    class ClazzViewHolder extends RecyclerView.ViewHolder {

        ClazzViewHolder(View itemView) {
            super(itemView);
        }
    }

    ClazzListRecyclerAdapter(@NonNull DiffUtil.ItemCallback<ClazzWithNumStudents>
                   diffCallback, Context context, Fragment fragment, ClazzListPresenter mPresenter) {
        super(diffCallback);
        theContext = context;
        theFragment = fragment;
        thePresenter = mPresenter;
    }

    @NonNull
    @Override
    public ClazzViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View clazzListItem =
                LayoutInflater.from(theContext).inflate(R.layout.item_clazzlist_clazz,
                        parent, false);
        return new ClazzViewHolder(clazzListItem);
    }

    /**
     * This method sets the elements after it has been obtained for that item'th position.
     *
     * @param holder The view holder
     * @param position The position of the item
     */
    @Override
    public void onBindViewHolder  (@NonNull ClazzViewHolder holder, int position) {
        ClazzWithNumStudents clazz = getItem(position);
        assert clazz != null;
        long attendancePercentage = (long) (clazz.getAttendanceAverage() * 100);
        String lastRecordedAttendance = "";
        ((TextView)holder.itemView.findViewById(R.id.item_clazzlist_clazz_title))
                .setText(clazz.getClazzName());
        String numStudentsText = clazz.getNumStudents() + " " + theFragment.getResources()
                .getText(R.string.students_literal).toString();
        String attendancePercentageText =
                attendancePercentage + "% " + theFragment.getText(R.string.attendance)
                        + " (" + theFragment.getText(R.string.last_recorded)
                        + " " + lastRecordedAttendance + ")";
        ((TextView)holder.itemView.findViewById(R.id.item_clazzlist_numstudents_text))
                .setText(numStudentsText);
        ((TextView)holder.itemView.findViewById(R.id.item_clazzlist_attendance_percentage))
                .setText(attendancePercentageText);
        holder.itemView.setOnClickListener((view) -> thePresenter.handleClickClazz(clazz));
        holder.itemView.findViewById(R.id.item_clazzlist_attendance_record_attendance_button)
                .setOnClickListener((view)-> thePresenter.handleClickClazzRecordAttendance(clazz));

        ImageView trafficLight =
                holder.itemView.findViewById(R.id.item_clazzlist_attendance_trafficlight);
        if(attendancePercentage > 75L){
            trafficLight.setColorFilter(ContextCompat.getColor(theContext, R.color.traffic_green));
        }else if(attendancePercentage > 50L){
            trafficLight.setColorFilter(ContextCompat.getColor(theContext, R.color.traffic_orange));
        }else{
            trafficLight.setColorFilter(ContextCompat.getColor(theContext, R.color.traffic_red));
        }
    }
}

File: app-android/src/main/java/com/ustadmobile/port/android/view/ClazzListRecyclerAdapter.kt
package com.ustadmobile.port.android.view

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.toughra.ustadmobile.databinding.ItemClazzListCardBinding
import com.ustadmobile.core.controller.ClazzListItemListener
import com.ustadmobile.core.controller.TerminologyKeys
import com.ustadmobile.core.util.safeParse
import com.ustadmobile.lib.db.entities.ClazzWithListDisplayDetails
import com.ustadmobile.port.android.view.util.SelectablePagedListAdapter
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import org.kodein.di.DI

class ClazzListRecyclerAdapter(var itemListener: ClazzListItemListener?, val di: DI)
    : SelectablePagedListAdapter<ClazzWithListDisplayDetails,
        ClazzListRecyclerAdapter.ClazzList2ViewHolder>(DIFF_CALLBACK) {

    class ClazzList2ViewHolder(val itemBinding: ItemClazzListCardBinding)
        : RecyclerView.ViewHolder(itemBinding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ClazzList2ViewHolder {
        val itemBinding = ItemClazzListCardBinding.inflate(LayoutInflater.from(parent.context), parent,
                false)
        return ClazzList2ViewHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: ClazzList2ViewHolder, position: Int) {
        val clazz = getItem(position)
        holder.itemBinding.clazz = clazz
        holder.itemView.tag = holder.itemBinding.clazz?.clazzUid
        holder.itemBinding.itemListener = itemListener
        val termMap: Map<String, String> =  clazz?.terminology?.ctTerminology?.let {
            safeParse(di, MapSerializer(String.serializer(), String.serializer()), "")
        } ?: mapOf()
        holder.itemBinding.teacherStudentCount = """${clazz?.numTeachers ?: 0} ${termMap[TerminologyKeys.TEACHERS_KEY]}, ${clazz?.numStudents ?: 0} ${termMap[TerminologyKeys.STUDENTS_KEY]}"""
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        itemListener = null
    }

    companion object {
        val DIFF_CALLBACK: DiffUtil.ItemCallback<ClazzWithListDisplayDetails> = object
            : DiffUtil.ItemCallback<ClazzWithListDisplayDetails>() {
            override fun areItemsTheSame(oldItem: ClazzWithListDisplayDetails,
                                         newItem: ClazzWithListDisplayDetails): Boolean {
                return oldItem.clazzUid == newItem.clazzUid
            }

            override fun areContentsTheSame(oldItem: ClazzWithListDisplayDetails,
                                            newItem: ClazzWithListDisplayDetails): Boolean {
                return oldItem.clazzName == newItem.clazzName &&
                        oldItem.numStudents == newItem.numStudents &&
                        oldItem.numTeachers == newItem.numTeachers &&
                        oldItem.clazzDesc == newItem.clazzDesc &&
                        oldItem.clazzActiveEnrolment?.clazzEnrolmentRole == newItem.clazzActiveEnrolment?.clazzEnrolmentRole &&
                        oldItem.attendanceAverage == newItem.attendanceAverage
            }
        }
    }

}